### PR TITLE
niLang/niCurl: Fetch request override from JavaScript code

### PR DESCRIPTION
**Pull request by JavierFerrerGomez**
## Description, Motivation and Context
The web client needs the option to implement a JavaScript function that overrides the call of niCurl fetch function. It can basically intercept the call and override it with anything it wants. One use case is to connect to a Crypto wallet. This function can handle call the wallet API and only return when all the transactions have been finalized. From the point of view of the niCURL client this all happens transparently.

To have access to all the emscripten environment this function can be implemented in the post_js_0.js file. The function module path is: Module.niCURL.handleFetchOverride

Example:
```
handleFetchOverride: function (request_url) {
  console.log("Module.niCURL: handleFetchOverride");
  return `
      {
        "status": "OK",
        "url": "http://example.com",
        "headers": {
          "Content-Type": "application/json",
          "Accept": "application/json"
        },
        "payload": "the actual request reply"
      }`
}
```
This function would just generate a json reply and return it to the client.

See the file post_js_0.js of this PR for more details.

* Link to issue: https://github.com/reddoordigital/reddoor/issues/575

## How Has This Been Tested?

For Linux or MacOS:
```
cd niLang/sources/niCurl
. hat emscripten
./ham all
```
How to test:
```
cd $WORK
simple-http-server regular .
```
Now open a web browser and go to this URL:
```
http://localhost:8123/niLang/bin/web-js/Test_niCURL_ra.html?niCURL.TraceFetch=1```
```

For Windows. There is not emscripten, so you can test only the standard niCURL:
How to Build
```
cd niLang/sources/niCURL
. hat
./ham Run_Test_niCURL
```

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Demos or Packages (adds examples or modifies a release package)
* [ ] Documentation (code docs, comments, or changes to the doc systems)
* [ ] Testing/Build (test coverage or the test/build subsystems themselves)

### CODE REVIEW

File: sources/emscripten/post_js_0.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    //     return `
+    //         {
+    //           "url": "http://example.com",
+    //           "headers": {
+    //             "Content-Type": "application/json",
+    //             "Accept": "application/json"
+    //           },
+    //           "payload": {
+    //             "status": "OK",
+    //             "data": "this will be send back to client in cml format"
```
Comment (prenaux) [inline sources/emscripten/post_js_0.js:22]: what's cml? shouldnt this be json?

File: sources/niCURL/src/CURL.cpp [hunk 2]
```diff
@@ -1401,6 +1408,98 @@ class cCURL : public cIUnknownImpl<iCURL>
...
+          dataFp,
+          apSink
+        );
+        // NOTE: this is kind of rude but this is a quite special case anyways...
+        request->_status = 500;
+        return request;
+      }
+    }
+    else {
+      niDebugFmt(("niCURL: Using emscripten fetch..."));
```
Comment (prenaux) [inline sources/niCURL/src/CURL.cpp:1500]: TRACE_FETCH instead of niDebugFmt?

File: sources/niCURL/src/CURL.cpp [hunk 3]
```diff
@@ -1401,6 +1408,98 @@ class cCURL : public cIUnknownImpl<iCURL>
                                       iFetchSink* apSink,
                                       const tStringCVec* apHeaders)
   {
+    if (_hasFetchOverride) {
```
Comment (prenaux) [inline sources/niCURL/src/CURL.cpp:1411]: This should move to its own function for clarity. Probably better to have _FetchWithOverride and _FetchEmscripten.

File: sources/niCURL/src/CURL.cpp [hunk 4]
```diff
@@ -1401,6 +1408,98 @@ class cCURL : public cIUnknownImpl<iCURL>
...
+      }
+      else {
+        TRACE_FETCH(("niCurl: handleFetchOverride exists but returns garbage. We return a 500 error."));
+        Ptr<iFile> dataFp = ni::CreateFileDynamicMemory(0,NULL);
+        niCheckIsOK(dataFp,NULL);
+        cString payload = cString(R"""(
+<payload status="ERROR">
+  <data message="Invalid handleFetchOverride"/>
+</payload>
+/>)""");
```
Comment (prenaux) [inline sources/niCURL/src/CURL.cpp:1484]: Extra `/>`

File: sources/niCURL/src/CURL.cpp [hunk 5]
```diff
@@ -1401,6 +1408,98 @@ class cCURL : public cIUnknownImpl<iCURL>
...
+          apSink
+        );
+        TRACE_FETCH(("... Fetch[%s]: Created reply request object", request->_url));
+        return request;
+      }
+      else {
+        TRACE_FETCH(("niCurl: handleFetchOverride exists but returns garbage. We return a 500 error."));
+        Ptr<iFile> dataFp = ni::CreateFileDynamicMemory(0,NULL);
+        niCheckIsOK(dataFp,NULL);
+        cString payload = cString(R"""(
```
Comment (prenaux) [inline sources/niCURL/src/CURL.cpp:1480]: We should return JSON for all web APIs.

File: sources/niCURL/src/CURL.cpp [hunk 6]
```diff
@@ -1401,6 +1408,98 @@ class cCURL : public cIUnknownImpl<iCURL>
...
                                       const tStringCVec* apHeaders)
   {
+    if (_hasFetchOverride) {
+      TRACE_FETCH(("niCURL: Using JS fetch override."));
+      cString retString = (const char*) EM_ASM_PTR({
+        console.log("niCURL: Fetching using Module.niCURL.handleFetchOverride");
+        let result = Module.niCURL.handleFetchOverride();
+        return allocateUTF8(result);
+      });
+      if (!retString.empty()) {
```
Comment (prenaux) [inline sources/niCURL/src/CURL.cpp:1418]: We need a way to indicate that we want the regular fetch because the override will only be for some URL.

**Review by prenaux — CHANGES_REQUESTED** [review]
See comments for details:
- The override to only hook some URLs. Maybe if it returns "undefined" it just uses the regular path.
- Please add test case in niCURL/tsrc.

File: sources/niCURL/src/CURL.cpp (hunk 2 above)
Comment (JavierFerrerGomez) [reply sources/niCURL/src/CURL.cpp]: Lapsus :P

**Comment by jamesrdd** [comment]
In the `handleFetchOverride()` javascript function, we need to optionally override some API endpoints while skipping the override for other endpoints. In order to achieve this, I think `handleFetchOverride()` needs to receive the URL being fetched as a parameter so we can add this custom override behavior only when fetching specific URLs.

File: sources/emscripten/post_js_0.js [hunk 7]
```diff
@@ -20,7 +20,7 @@ niExtensions = {
     //    - In the case of "ERROR" status it contains a description of the error.
     //    - In the case of "SKIP" status it contains an empty string.
     // Example:
-    //   handleFetchOverride: function () {
+    //   handleFetchOverride: function (request_url) {
```
Comment (prenaux) [inline sources/emscripten/post_js_0.js:23]: `request_url` -> `aRequestUrl`

File: sources/niCURL/tsrc/Test_Fetch.cpp [hunk 8]
```diff
@@ -42,8 +42,8 @@ TEST_FIXTURE(FCURLFetch,OverrideFetchSkip) {
   emscripten_run_script(R"""({
     niExtensions = {
       niCURL: {
-        handleFetchOverride: function () {
-          console.log("Module.niCURL: handleFetchOverride");
+        handleFetchOverride: function (request_url) {
```
Comment (prenaux) [inline sources/niCURL/tsrc/Test_Fetch.cpp:45]: `request_url` -> `aRequestUrl`

File: sources/niCURL/tsrc/Test_Fetch.cpp [hunk 9]
```diff
@@ -101,8 +101,8 @@ TEST_FIXTURE(FCURLFetch,OverrideFetchError) {
...
```
Comment (prenaux) [inline sources/niCURL/tsrc/Test_Fetch.cpp:104]: `request_url` -> `aRequestUrl`

File: sources/niCURL/tsrc/Test_Fetch.cpp [hunk 10]
```diff
@@ -257,8 +257,8 @@ TEST_FIXTURE(FCURLFetch,GetJson) {
...
```
Comment (prenaux) [inline sources/niCURL/tsrc/Test_Fetch.cpp:260]: `request_url` -> `aRequestUrl`

**Review by prenaux — APPROVED** [review]
Ideally rename the parameters before merging. See comments.

**Pull request merged**